### PR TITLE
[DI] use dirname() when possible

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

On second look, I think this works :)

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 1]
```diff
@@ -1350,12 +1341,11 @@ private function addInlineRequires(?array &$preload): string
         foreach ($lineage as $file) {
             if (!isset($this->inlinedRequires[$file])) {
                 $this->inlinedRequires[$file] = true;
-                $file = preg_replace('#^\\$this->targetDirs\[(\d++)\]#', sprintf('\dirname(__DIR__, %d + $1)', $this->asFiles), $file);
                 $code .= sprintf("\n            include_once %s;", $file);
             }
         }
 
-        return $code ? sprintf("\n        \$this->privates['service_container'] = static function () {%s\n        };\n", $code) : '';
+        return $code ? sprintf("\n        \$this->privates['service_container'] = function () {%s\n        };\n", $code) : '';
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: this seems unrelated, right ?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: it is: we didn't generate `$this` anymore inside, but that's now back (and legit)

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: https://github.com/symfony/symfony/pull/33529/files#diff-edd219a3021114261638a490a9e00ae9L93-R91

Comment (stof) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: hmm, so this was broken previously when changing the container dir ?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: possibly yes

**Review by stof — APPROVED** [review]

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 2]
```diff
@@ -306,7 +312,7 @@ public function dump(array $options = [])
 
 use Symfony\Component\DependencyInjection\Dumper\Preloader;
 
-require dirname(__DIR__, 3).'/vendor/autoload.php';
+require $autoloadFile;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: now with proper detection of `autoload.php` location

Comment (stof) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Do we have any test covering the dumping of this file ?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: yes; not precisely this line but the `$classes[] = ...` one below
